Url and path pattern matching refactor

alvinjohnsonso opened the pull request:
- Removed previous matchers
- Added modules for Url and Path Pattern matching
- Url Pattern Matching matches a single url to multiple patterns
- Path Pattern Matching matches a single path chunks to multiple path chunks
- Added helper files for Path and Url

alvinjohnsonso committed:
Url and path pattern matching refactor
- Removed previous matchers
- Added modules for Url and Path Pattern matching
- Url Pattern Matching matches a single url to multiple patterns
- Path Pattern Matching matches a single path chunks to multiple path chunks
- Added helper files for Path and Url

alvinjohnsonso committed:
Added handlers for
- patterns that are similar to current path but has an additional wildcard at the end
- patterns that has leading wildcard and is longer than the current path
- patterns that are shorter than the current path

alvinjohnsonso committed:
- Changed UrlHelper parse_pattern_url to parse_url and have path return default '/' value
- Reorganized and added test scenarios for url pattern matcher
- Updated PathHelper get_chunks method to return empty string array if path provided is only a slash

alvinjohnsonso committed:
- Updated get_chunks to return empty array if no paths are provided
- Simplified if else statements for path matching and url checking

alvinjohnsonso committed:
added .gitattributes file to ignore .github folder on export

alvinjohnsonso committed:
added return false for handler of empty current path chunks if the pattern path is not only a wildcard

alvinjohnsonso committed:
added test for matching empty current path to path with trailing wildcard pattern

alvinjohnsonso committed:
used regex for protocol matching

alvinjohnsonso committed:
Updated php docs of each functions available and added README content

alvinjohnsonso committed:
Added changelog

alvinjohnsonso committed:
version bumped to 2.0.0

GeekOfAges commented on lib/Helpers/UrlHelper.php:
```diff
@@ -0,0 +1,46 @@
+		$has_protocol = preg_match(PROTOCOL_REGEX, $pattern) === 1;
+
+		if ($is_path === false && $has_protocol === false) {
+			// add http:// in front of the string
```
Maybe note down reason as to why this is important (parsing)

GeekOfAges commented on lib/Modules/PathPatternMatcher.php:
```diff
@@ -0,0 +1,169 @@
+			//   pattern - /path/to/somewhere/*
+			//
+			// check the next pattern chunk if it's a wildcard
+			if ($i + 1 === $current_path_len && isset($pattern_path_chunks[$i + 1])) {
```
Not entirely sure this is necessary since `/path/other` and `/path` are 2 distinct paths.
This should be handled by the user by providing 2 patterns, one for strict `/path` and one for `/path/*`

alvinjohnsonso replied:
Ahh, the reason I added this part is because of how the current behavior of the WordPress plugin.

GeekOfAges commented on lib/Modules/PathPatternMatcher.php:
```diff
@@ -0,0 +1,169 @@
+	 * Iterates over the current and pattern path chunks and matches them
+	 */
+	private static function match_chunks($current_path_chunks, $pattern_path_chunks) {
+		$wildcard_loc = PathHelper::get_wildcard_location_by_chunks($pattern_path_chunks);
```
We're getting the location in here and in `match` which seems wasteful. Could provide `stop_on_wildcard` parameter as `$wildcard_loc === WildcardLocation::MIDDLE`

alvinjohnsonso committed:
- Added phpdocs for function parameters
- Added PathPattern Model
- Added handler for patterns with leading slash and wildcard
- Added handler for patterns with trailing wildcard and slash
- Added handler for patterns with leading and middle wildcards
- Added handler for patterns with trailing and middle wildcards

alvinjohnsonso committed:
added path pattern model tests

alvinjohnsonso committed:
Merge pull request #5 from alvinjohnsonso/path-pattern-additional-matching

Path pattern additional matching
Added handler for following patterns:
- leading slash and wildcard
- trailing wildcard and slash
- leading and middle wildcards
- trailing and middle wildcards

alvinjohnsonso committed:
Merge branch 'feature/url-and-path-matching-modules' of github.com:alvinjohnsonso/tawk-url-utils into url-and-path-matching-modules-docs

alvinjohnsonso committed:
- Added new valid formats for multiple wildcard patterns
- Aligned php doc param descriptions

alvinjohnsonso committed:
Merge pull request #4 from alvinjohnsonso/url-and-path-matching-modules-docs

Url and path matching modules documentation
- Updated PHP doc for each function available
- Added @small annotation to test cases to identify that the unit test should be a small unit test.
- Updated README.md for module usage
- Added CHANGELOG.md
